cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -0,0 +1,550 @@
+                                  Map<Method, AnnotationDetail> methodAnnotations) {
+      super(Opcodes.ASM5, cw);
+      this.className = className;
+      this.classType = Type.getObjectType(className.replace(".", "/"));
+      this.fieldDetails = fieldDetails;
+      this.methodAnnotations = methodAnnotations;
+      this.authenticationContextFieldName = generateUniqueFieldName(AUTHENTICATION_CONTEXT_FIELD_NAME);
+      this.authorizationEnforcerFieldName = generateUniqueFieldName(AUTHORIZATION_ENFORCER_FIELD_NAME);
+    }
+
+    @Override
+    public MethodVisitor visitMethod(int access, final String methodName, final String methodDesc, String signature,
+                                     String[] exceptions) {
+      MethodVisitor mv = super.visitMethod(access, methodName, methodDesc, signature, exceptions);
+      // From the first pass we know the methods which has AuthEnforce annotation and needs to be rewritten
+      // if this method was not identified earlier as marked with AuthEnforce annotation just skip it
+      if (!methodAnnotations.containsKey(new Method(methodName, methodDesc))) {
+        return mv;
+      }
+      return new AdviceAdapter(Opcodes.ASM5, mv, access, methodName, methodDesc) {
+        @Override
+        protected void onMethodEnter() {
+          // Get the annotation details which was collected by visiting annotations during the first pass
+          AnnotationDetail annotationDetail = methodAnnotations.get(new Method(methodName, methodDesc));
+
+          LOG.trace("AuthEnforce annotation found in class {} on method {}. Authorization enforcement command will " +
+                      "be generated for Entities: {}, enforceOn: {}, actions: {}.", className, methodName,
+                    annotationDetail.getEntities(), annotationDetail.getEnforceOn(),
+                    annotationDetail.getActions());
+
+          // TODO: Remove this one we support AuthEnforce with multiple string parts
+          Preconditions.checkArgument(annotationDetail.getEntities().size() == 1,
+                                      "Currently Authorization annotation is only supported for EntityId from " +
+                                        "method parameter.");
+
+          // do class rewrite to generate the call to
+          // AuthEnforceUtil#enforce(AuthorizationEnforcer, EntityId, AuthenticationContext, Set)
+
+          // this.authorizationEnforcer
+          loadThis();
+          getField(classType, authorizationEnforcerFieldName, AUTHORIZATION_ENFORCER_TYPE);
+
+          // push the parameters of method call on to the stack
+
+          // pushed the entity id
+          //TODO: This is because currently we don't support multi-part
+          String name = annotationDetail.getEntities().get(0);
+          if (name.startsWith("this.")) { // if the entity name was specified with this. then we only look for it in
```

Comment:
This logic could be simpler if we memorize the decision in the annotation detail, since in the first pass we already performed the validation logic, which means we already knew whether it is a field or a method param. We can even strip the `this.` there.